### PR TITLE
Implement DeepDDI model

### LINKED ISSUE #2: Add the DeepDDI model
Dear @hzcheney,

- Please read the paper first. It is [here](https://www.pnas.org/content/115/18/E4304).
- After that read the contributing guidelines.
-   If there is an existing open source version of the model please take a look.
-   ChemicalX is built on top of [PyTorch 1.10.](https://pytorch.org/) and [torchdrug](https://torchdrug.ai/).
- A similar model is `` which uses``  to generate drug representations. Take a look at the layer definition [here]().
- The library heavily builds on top on [torchdrug](https://torchdrug.ai/) and molecules in batches are PackedGraphs.
- There is already a model class under `./chemicalx/models/`
- Context features, drug level features and labels are all FloatTensors.
- Look at the examples and tests under `./examples/` and `./tests/`.
- Add auxiliary layers as you see fit - please document these, add tests and add these layers to the main readme.md if needed.
- Add typing to the initialisation and forward pass.
- Non data dependent hyperparameters should have default values. 
- Please add tests under `./tests/` and make sure that your model/layer is tested with real data.
- Write an example under `./examples/`. What is the AUC on the test set? Is it reasonable?

### PR DESCRIPTION
**Pull request by hzcheney**
Closes #2
 
- [x] Unit tests provided for these changes
- [ ] Documentation and docstrings added for these changes

## Changes 

* _add DeepDDI model_
* _add DeepDDI examples_
*  _add DeepDDI  testcase_

### CODE REVIEW

File: chemicalx/models/deepddi.py [hunk 1]
```diff
@@ -1,14 +1,90 @@
...
+            torch.nn.ReLU(),
+            torch.nn.BatchNorm1d(
+                num_features=hidden_channels,
+                affine=True,
+                momentum=None
+            ),
+            torch.nn.ReLU()
+        ])
+        for _ in range(hidden_layers_num - 1):
+            dnn.extend([
```
Comment (cthoyt) [inline chemicalx/models/deepddi.py:48]: just append the element instead of extending on a list

File: chemicalx/models/deepddi.py [hunk 2]
```diff
@@ -1,14 +1,90 @@
...
+        :param drug_channels: The number of drug features.
+        :param hidden_channels: The number of hidden layer neurons.
+        :param hidden_layers_num: The number of hidden layers.
+        :param out_channels: The number of output channels.
+        """
+
+        super(DeepDDI, self).__init__()
+        assert hidden_layers_num > 1
+        dnn = []
+        dnn.extend([
```
Comment (cthoyt) [inline chemicalx/models/deepddi.py:37]: why define the list then immediately extend it? just define the base elements directly

**Comment by cthoyt** [comment]
Please run `tox` locally to make sure that all tests are passing before we give a further review

File: chemicalx/models/deepddi.py (hunk 1 above)
Comment (hzcheney) [reply chemicalx/models/deepddi.py:48]: Did you mean append each element separately? :-)

**Comment by benedekrozemberczki** [comment]
Hi there @hzcheney, do you plan to update this PR?

**Comment by hzcheney** [comment]
> Hi there @hzcheney, do you plan to update this PR?

Yes! I will update this PR ASAP.

File: chemicalx/models/deepddi.py (hunk 2 above)
Comment (cthoyt) [reply chemicalx/models/deepddi.py:37]: e.g., 
```python
dnn = [
            torch.nn.Linear(drug_channels * 2, hidden_channels),
            torch.nn.ReLU(),
            torch.nn.BatchNorm1d(
                num_features=hidden_channels,
                affine=True,
                momentum=None
            ),
            torch.nn.ReLU()
]
```

File: chemicalx/models/deepddi.py (hunk 1 above)
Comment (cthoyt) [reply chemicalx/models/deepddi.py:48]: Sorry I misread this part in the github viewer, on a second way the way you did it was very nice!

**Comment by benedekrozemberczki** [comment]
The formatting still fails.

File: chemicalx/models/deepddi.py (hunk 1 above)
Comment (hzcheney) [reply chemicalx/models/deepddi.py:48]: That's OK! Thank you! :-)

**Comment by benedekrozemberczki** [comment]
These are the errors:

./tests/unit/test_models.py:61:59: BLK100 Black would make changes.
./examples/deepddi_examples.py:22:10: BLK100 Black would make changes.
./examples/epgcnds_examples.py:15:55: BLK100 Black would make changes.

**Comment by hzcheney** [comment]
I am working on it, it's weird that I run `tox` locally well.

**Comment by hzcheney** [comment]
Please help! I have no idea why these errors still happened:
```
./tests/unit/test_models.py:61:59: BLK100 Black would make changes.
./examples/epgcnds_examples.py:15:55: BLK100 Black would make changes.
```
Even I don't change the `epgcnds_examples.py` file.
:-)

**Comment by cthoyt** [comment]
Yesterday, `black` made a new release that will cause some changes. The options are either:
1. Wait for a new PR where I make these changes
2. Pin black (terrible idea! Do not do this!)
3. Commit the changes black makes when you run `tox`

**Comment by hzcheney** [comment]
> Yesterday, `black` made a new release that will cause some changes. The options are either:
> 
> 1. Wait for a new PR where I make these changes
> 2. Pin black (terrible idea! Do not do this!)
> 3. Commit the changes black makes when you run `tox`

Thank you! It's clear to me now.

File: chemicalx/models/deepddi.py [hunk 3]
```diff
@@ -1,14 +1,78 @@
...
+        """
+        Run a forward pass of the DeepDDI model.
+
+        Args:
+            drug_features_left (torch.FloatTensor): A matrix of head drug features.
+            drug_features_right (torch.FloatTensor): A matrix of tail drug features.
+        Returns:
+            hidden (torch.FloatTensor): A column vector of predicted interaction scores.
+        """
+        input_feature = torch.cat([drug_features_left, drug_features_right], 1)
```
Comment (cthoyt) [inline chemicalx/models/deepddi.py:76]: can you make the kwarg for the dimension explicit please

File: examples/deepddi_examples.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+from chemicalx import pipeline
+from chemicalx.data import DrugbankDDI
+from chemicalx.models import DeepDDI
+
+
+def main():
+    """Train and evaluate the DeepSynergy model."""
+    dataset = DrugbankDDI()
+    model = DeepDDI(drug_channels=dataset.drug_channels, hidden_layers_num=2)
```
Comment (cthoyt) [inline examples/deepddi_examples.py:11]: why does the number of hidden layers in the example differ from the chosen default?

Comment (hzcheney) [reply examples/deepddi_examples.py:11]: It is intended to make the training process faster.

Comment (cthoyt) [reply examples/deepddi_examples.py:11]: I think we want the examples to show configurations that get good results. Does this model take a long time to run on this dataset?

File: chemicalx/models/deepddi.py (hunk 3 above)
Comment (hzcheney) [reply chemicalx/models/deepddi.py:76]: It just combined the two drug features and I will update this PR later due to the Spring Festival. :-)

File: examples/deepddi_examples.py (hunk 4 above)
Comment (hzcheney) [reply examples/deepddi_examples.py:11]: It runs slowly on just CPUs. You're right and I will change it to the default settings as same as the original paper.

**Comment by benedekrozemberczki** [comment]
Hi there @hzcheney, I suggest updating black on your side. Do you want to finish the PR or we can overtake from you?

**Review by cthoyt — APPROVED** [review]

File: chemicalx/models/deepddi.py (hunk 3 above)
Comment (cthoyt) [reply chemicalx/models/deepddi.py]: Done in c16f927

File: chemicalx/models/deepddi.py [hunk 5]
```diff
@@ -1,14 +1,90 @@
...
 
     .. seealso:: https://github.com/AstraZeneca/chemicalx/issues/2
     """
+
+    def __init__(
+        self,
+        *,
+        drug_channels: int,
+        hidden_channels: int = 2048,
+        hidden_layers_num: int = 9,
```
[the remark this answers is not in the document]
Comment (benedekrozemberczki) [reply chemicalx/models/deepddi.py:29]: @cthoyt  can we have 4 layer with 32 hidden channels?

**Review by benedekrozemberczki — CHANGES_REQUESTED** [review]
@cthoyt can you make these changes?

**Review by benedekrozemberczki — APPROVED** [review]
